From 4b8caed44f21bfac807f574ba979c2c4fcf27b27 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Wed, 13 Jun 2018 10:28:29 +0900
Subject: [PATCH 314/909] drm/brdige: dumb-vga-dac: Add dot clock valid check

commit 59c65194d7306634450c5cabe7f615e4c96a42b8 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

This patch register the mode_valid function and check the dot clock.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/dumb-vga-dac.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/gpu/drm/bridge/dumb-vga-dac.c b/drivers/gpu/drm/bridge/dumb-vga-dac.c
index 8590e73590b2..6433a4b5461b 100644
--- a/drivers/gpu/drm/bridge/dumb-vga-dac.c
+++ b/drivers/gpu/drm/bridge/dumb-vga-dac.c
@@ -30,6 +30,7 @@ struct dumb_vga {
 	bool			ddc_flag;
 	u32			width;
 	u32			height;
+	u32			max_pixclk;
 };
 
 static inline struct dumb_vga *
@@ -77,8 +78,24 @@ static int dumb_vga_get_modes(struct drm_connector *connector)
 	return ret;
 }
 
+static enum drm_mode_status
+dumb_vga_mode_valid(struct drm_connector *connector,
+		    struct drm_display_mode *mode)
+{
+	struct dumb_vga *vga = drm_connector_to_dumb_vga(connector);
+
+	if (!vga->max_pixclk)
+		return MODE_OK;
+
+	if (mode->clock > vga->max_pixclk)
+		return MODE_BAD;
+
+	return MODE_OK;
+}
+
 static const struct drm_connector_helper_funcs dumb_vga_con_helper_funcs = {
 	.get_modes	= dumb_vga_get_modes,
+	.mode_valid	= dumb_vga_mode_valid,
 };
 
 static enum drm_connector_status
@@ -187,6 +204,8 @@ static struct i2c_adapter *dumb_vga_retrieve_ddc(struct device *dev)
 		vga->height = 768;
 	}
 
+	of_property_read_u32(remote, "max-pixelclock", &vga->max_pixclk);
+
 	phandle = of_parse_phandle(remote, "ddc-i2c-bus", 0);
 	of_node_put(remote);
 	if (!phandle)
-- 
2.17.1

